arch/arc/core/thread.c

```diff
@@ -64,14 +64,41 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
 	char *stackEnd = pStackMem + stackSize;
 	struct init_stack_frame *pInitCtx;
-
+#if CONFIG_USERSPACE
+	/* for kernel thread, the privilege stack is merged into thread stack */
+	if (!(options & K_USER)) {
+	/* if MPU_STACK_GUARD is enabled, reserve the the stack area
+	 * |---------------------|----------------|
+	 * |                     |(MPU STACK AREA)|
+	 * |			 |----------------|
+	 * |  user stack         | kernel thread  |
+	 * |---------------------| stack          |
+	 * |  privilege stack    |                |
+	 * ----------------------------------------
+	 */
+#ifdef CONFIG_MPU_STACK_GUARD
+		pStackmem += STACK_GUARD_SIZE;
+		stackSize = stackSize + CONFIG_PRIVILEGED_STACK_SIZE
+			 - STACK_GUARD_SIZE;
+#endif
+		stackEnd += CONFIG_PRIVILEGED_STACK_SIZE;
+		stackSize += CONFIG_PRIVILEGED_STACK_SIZE;
+	}
+#endif
 	_new_thread_init(thread, pStackMem, stackSize, priority, options);
 
 	/* carve the thread entry struct from the "base" of the stack */
 	pInitCtx = (struct init_stack_frame *)(STACK_ROUND_DOWN(stackEnd) -
 				       sizeof(struct init_stack_frame));
-
+#if CONFIG_USERSPACE
+	if (options & K_USER) {
+		pInitCtx->pc = ((u32_t)_arch_user_mode_enter);
```

Comment:
This initialization depending on thread creation option doesn't work when instead of at init, a thread drops privilege level at runtime. Simplest would be to always initialize both privilege and userstack areas, but that results in always fixed stack for kernel mode, even when never dropping to user. Should some option flag be added to the create call to indicate the three options: always kernel, drop to user, and always user?

Reply:
> that results in always fixed stack for kernel mode, even when never dropping to user

That's how we are currently doing it on x86.

> Should some option flag be added to the create call to indicate the three options: always kernel, drop to user, and always user?

We would have to have extra stack declaration macros as well such that you could declare thread stacks that don't have memory reserved for a kernel stack.

I suggest we punt on this for now, but maybe file a github bug about it looking into it further.